### PR TITLE
Replace str.format with f-strings

### PR DESCRIPTION
**Pull request by Armavica**
## Related Issue

- [ ] Closes #
- [ ] Related to #

## Checklist

- [ ] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [x] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [ ] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [x] Maintenance
- [ ] Other (please specify):

### CODE REVIEW

File: pytensor/link/c/cmodule.py [hunk 1]
```diff
@@ -1961,14 +1961,14 @@ def _try_flags(
...
-            """
+            f"""
         {preamble}
         int main(int argc, char** argv)
         {{
             {body}
             return 0;
         }}
-        """.format(**locals())
+        """
```
Comment (michaelosthege) [inline pytensor/link/c/cmodule.py:1971]: Shouldn't `**locals` expand to a dictionary of all function parameters, not just `preamble`?

File: pytensor/scalar/math.py [hunk 2]
```diff
@@ -621,8 +621,8 @@ def c_code(self, node, name, inp, out, sub):
         (z,) = out
         if node.inputs[0].type in float_types:
             dtype = "npy_" + node.outputs[0].dtype
-            return """{z} =
-                ({dtype}) 1 - GammaP({k}/2., {x}/2.);""".format(**locals())
+            return f"""{z} =
+                ({dtype}) 1 - GammaP({k}/2., {x}/2.);"""
```
Comment (michaelosthege) [inline pytensor/scalar/math.py:625]: Again with `**locals`, also below

**Review by michaelosthege — COMMENTED** [review]
this is quite time-consuming to review. How much of is was automated?

In some places there were `**locals` or `locals()`, and I don't have much experience with those. (Which is a good thing.)
But can you double-check these places?

**Comment by Armavica** [comment]
> this is quite time-consuming to review. How much of is was automated?

None, everything was manual…

File: pytensor/link/c/cmodule.py (hunk 1 above)
Comment (Armavica) [reply pytensor/link/c/cmodule.py:1971]: `locals()` is a dictionary of all the variables in scope, so it expands here to `preamble` and `body`, but also to others. However, `format` doesn't use its unnecessary arguments.
``` python
assert "{abc}".format(abc="X", xyz="Y") == "X"
```

File: pytensor/compile/debugmode.py [hunk 3]
```diff
@@ -1728,15 +1724,11 @@ def f():
...
-                                + str(reason[0][1])
-                                + "\n"
-                                + reason[0][2]
-                                + "\n\nThe original exception: \n"
-                                + str(e)
+                                f"An optimization (probably {opt}) inserted an "
+                                "apply node that raise an error.\n"
+                                "The information we have about this optimization is:"
+                                f"{reason[0][1]}\n{reason[0][2]}\n"
+                                "\nThe original exception: \n{e}"
```
Comment (michaelosthege) [inline pytensor/compile/debugmode.py:1731]:
```suggestion
                                f"\nThe original exception: \n{e}"
```

File: pytensor/compile/debugmode.py [hunk 4]
```diff
@@ -1617,14 +1617,10 @@ def f():
...
-                                "optimizations is:"
-                                + str(reason[0][1])
-                                + "\n"
-                                + reason[0][2]
-                                + "\n\nThe original exception: \n"
-                                + str(e)
+                                "apply node that raise an error.\n"
+                                "The information we have about this optimization is:"
+                                f"{reason[0][1]}\n{reason[0][2]}\n"
+                                "\nThe original exception: \n{e}"
```
Comment (michaelosthege) [inline pytensor/compile/debugmode.py:1623]:
```suggestion
                                f"\nThe original exception: \n{e}"
```

File: pytensor/sparse/rewriting.py [hunk 5]
```diff
@@ -179,7 +179,7 @@ def c_code(self, node, name, inputs, outputs, sub):
                   }}
                  }}
                 }}
-             """.format(**dict(locals(), **sub))
+             """
```
Comment (michaelosthege) [inline pytensor/sparse/rewriting.py:182]: I think the contents of `sub` are not accounted for

Below you did
```python
fail = sub["fail"]
rval = f"""
```

File: pytensor/tensor/blas.py [hunk 6]
```diff
@@ -772,27 +772,15 @@ def build_gemm_call(self):
...
-            )
+        return (
+            f"{self.declare_NS}{self.check_xyz_rank2}{setup_z_Nz_Sz}"
+            f"{self.check_xyz_double_or_float}{self.check_ab_double_or_float}"
+            f"{self.broadcast_xy}{self.check_dims}{self.check_strides}"
+            f"{self.encode_strides_in_unit}{self.compute_strides}"
+            f"{self.begin_switch_typenum}{self.case_float}"
+            f"{self.case_float_ab_constants}{self.case_float_gemm}"
+            f"{self.case_double}{self.case_double_ab_constants}"
+            f"{self.case_double_gemm}{self.end_switch_typenum}"
```
Comment (michaelosthege) [inline pytensor/tensor/blas.py:783]: The previous version was much easier to read

File: pytensor/tensor/conv/abstract_conv.py [hunk 7]
```diff
@@ -633,7 +633,7 @@ def border_mode_to_pad(mode, convdim, kshp):
             ):
                 if len(m) != 2:
                     raise NotImplementedError(
-                        f"Asymmetric padding not implemented for {len(m)}d"
+                        f"Asymmetric padding not implemented for {len(m)}"
```
Comment (michaelosthege) [inline pytensor/tensor/conv/abstract_conv.py:636]: Was the `d` referring to dimensions? As in `m dimensions`?

File: pytensor/tensor/shape.py [hunk 8]
```diff
@@ -316,7 +316,7 @@ def grad(self, inp, grads):
                 op=self,
                 x_pos=0,
                 x=inp[0],
-                comment=("No gradient for the shape of a matrix " "is implemented."),
+                comment=("No gradient for the shape of a matrix is implemented."),
```
Comment (michaelosthege) [inline pytensor/tensor/shape.py:319]:
```suggestion
                comment="No gradient for the shape of a matrix is implemented.",
```

**Comment by michaelosthege** [comment]
> None, everything was manual…

😵

File: pytensor/tensor/blas.py (hunk 6 above)
Comment (ricardoV94) [reply pytensor/tensor/blas.py:783]: agreed

File: pytensor/sparse/rewriting.py (hunk 5 above)
Comment (Armavica) [reply pytensor/sparse/rewriting.py:182]: I think `sub` was superfluous here: all the patterns in this f-string are already accounted by local variables

File: pytensor/compile/debugmode.py (hunk 4 above)
Comment (Armavica) [reply pytensor/compile/debugmode.py:1623]: Great catch!

File: pytensor/tensor/blas.py (hunk 6 above)
Comment (Armavica) [reply pytensor/tensor/blas.py:783]: agreed xD

File: pytensor/tensor/conv/abstract_conv.py (hunk 7 above)
Comment (Armavica) [reply pytensor/tensor/conv/abstract_conv.py:636]: Ah yes indeed, great catch again. I was mixed up with percent formatting where `d` stands for `double`.

**Comment by Armavica** [comment]
@michaelosthege Thank you very much for your thorough review! I answered your comments and fixed the errors that you pointed out.

**Review by michaelosthege — APPROVED** [review]

**Pull request merged**